### ISSUE TITLE
sasl anonymous support in twisted.words.protocol.jabber

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4067]
|<img alt="twonds's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| twonds reported|
|-|-|
|Trac ID|trac#4067|
|Type|enhancement|
|Created|2009-10-29 13:43:19Z|
|Branch|https://github.com/twisted/twisted/tree/xmpp-sasl-anonymous-4067|

There are other SASL authentication methods supported in t.w.p.jabber but not sasl anonymous.

There is a start of some code on github

http://github.com/twonds/twisted_helper/blob/master/anon.py

It needs to be moved into twisted and tests written etc. 

I will try and upload the patches to this ticket. 

Attachments:

* [anon.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f07/f07719b68ead188c19a737e1870f2594cd05903b/0eb2b4c7f68e6711b4d5bd929d2fb8cf911cbf3f.diff) (3151 bytes) - added by twonds on 2009-10-29 14:42:43Z - patch for anonymous sasl support and small test
* [anon1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f07/f07719b68ead188c19a737e1870f2594cd05903b/c8a775920815f1c7a7912c2574f12e330c6b8420.diff) (9167 bytes) - added by twonds on 2009-10-29 18:37:59Z - fixes review comments

<details><summary>Searchable metadata</summary>

```
trac-id__4067 4067
type__enhancement enhancement
reporter__twonds twonds
priority__normal normal
milestone__ 
branch__branches_xmpp_sasl_anonymous_4067 branches/xmpp-sasl-anonymous-4067
branch_author__twonds__ralphm twonds, ralphm
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1256823799000000 1256823799000000
changetime__1257086479000000 1257086479000000
version__None None
owner__ 
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm commented|
|-|-|

For clarity, the implementation in `anon.py` looks more complicated than the actual changes to Words. I estimate no more than 20 lines of code. I talked to twonds and offered to review the code when there are tests.

**Comment by twisted-trac** [comment]
|<img alt="twonds's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|twonds set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="twonds's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|twonds set owner to @ralphm<br>twonds set status to `new`|
|-|-|

the patch is against trunk

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set owner to twonds|
|-|-|

Thanks for this simple and straightforward patch.

 * `SASLInitiatingInitializer.setMechanism` currently has no tests. Your patch doesn't change that, and subsequently fails to show the mistakes in:
```python
raise sasl.SASLNoAccetableMechanisms()
```
  These are:
   * We are in the `sasl` module, so the prefix can be dropped.
   * The typo in the name of the exception class.
 Could you include tests for this method, preferably for all possible paths, but at least the ones for the cases associated with the ANONYMOUS mechanism.
 * Docstrings in Twisted have the summary on a line of its own. Please adjust `Anonymous` for that (see `Plain`).
 * `AnonymousTest` does not have a docstring, please refer to the class the test is for, using the `L{}` epytext construct.

**Comment by twisted-trac** [comment]
|<img alt="twonds's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|twonds set owner to @ralphm|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm commented|
|-|-|

(In [27388]) Branching to 'xmpp-sasl-anonymous-4067'

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm commented|
|-|-|

(In [27391]) Adjust copyright statement, and spacing, docs, tests to coding standard.

This also splits and simplifies the test for the unacceptable SASL mechanism
cases.

Re [#4067](https://github.com/twisted/twisted/issues/4067).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_8"></a>|@ralphm set status to `assigned`|
|-|-|

This looks great now. I cleaned up some stuff (see comment:7). Thanks.

I'm proceeding to merge this to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm set status to `closed`|
|-|-|

(In [27392]) Add ANONYMOUS SASL support for XMPP clients.

Author: twonds.
Reviewer: ralphm.
Fixes [#4067](https://github.com/twisted/twisted/issues/4067).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
